block/pending_blocks.go

```diff
@@ -1,64 +1,112 @@
 package block
 
 import (
-	"sort"
-	"sync"
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"sync/atomic"
 
+	ds "github.com/ipfs/go-datastore"
+
+	"github.com/rollkit/rollkit/store"
+	"github.com/rollkit/rollkit/third_party/log"
 	"github.com/rollkit/rollkit/types"
 )
 
+// LastSubmittedHeightKey is the key used for persisting the last submitted height in store.
+const LastSubmittedHeightKey = "last submitted"
+
 // PendingBlocks maintains blocks that need to be published to DA layer
+//
+// Important assertions:
+// - blocks are safely stored in database before submission to DA
+// - blocks are always pushed to DA in order (by height)
+// - DA submission of multiple blocks is atomic - it's impossible to submit only part of a batch
+//
+// lastSubmittedHeight is updated only after receiving confirmation from DA.
+// Worst case scenario is when blocks was successfully submitted to DA, but confirmation was not received (e.g. node was
+// restarted, networking issue occurred). In this case blocks are re-submitted to DA (it's extra cost).
+// rollkit is able to skip duplicate blocks so this shouldn't affect full nodes.
+// TODO(tzdybal): we shouldn't try to push all pending blocks at once; this should depend on max blob size
 type PendingBlocks struct {
-	pendingBlocks map[uint64]*types.Block
-	mtx           *sync.RWMutex
+	store  store.Store
+	logger log.Logger
+
+	// lastSubmittedHeight holds information about last block successfully submitted to DA
+	lastSubmittedHeight atomic.Uint64
 }
 
 // NewPendingBlocks returns a new PendingBlocks struct
-func NewPendingBlocks() *PendingBlocks {
+func NewPendingBlocks(store store.Store, logger log.Logger) *PendingBlocks {
 	return &PendingBlocks{
-		pendingBlocks: make(map[uint64]*types.Block),
-		mtx:           new(sync.RWMutex),
+		store:  store,
+		logger: logger,
 	}
 }
 
 // getPendingBlocks returns a sorted slice of pending blocks
 // that need to be published to DA layer in order of block height
-func (pb *PendingBlocks) getPendingBlocks() []*types.Block {
-	blocks := copyBlocks(pb)
-	sort.Slice(blocks, func(i, j int) bool {
-		return blocks[i].Height() < blocks[j].Height()
-	})
-	return blocks
-}
+func (pb *PendingBlocks) getPendingBlocks(ctx context.Context) ([]*types.Block, error) {
+	lastSubmitted := pb.lastSubmittedHeight.Load()
+	if lastSubmitted == 0 {
+		err := pb.loadFromStore(ctx)
+		if err != nil {
+			return nil, err
+		}
+		lastSubmitted = pb.lastSubmittedHeight.Load()
+	}
+	height := pb.store.Height()
+
+	if lastSubmitted == height {
+		return nil, nil
+	}
+	if lastSubmitted > height {
+		panic(fmt.Sprintf("height of last block submitted to DA (%d) is greater than height of last block (%d)",
```

Comment:
I also don't like panics.
But this should never happen. If it happens data is already corrupted - last block height was updated "manually", or some blocks were removed from the store, or there is some IO error, or something.

I'll think about more gentle "shutdown" procedure.

Reply:
I'm also find turning this into a follow up since there might be other places in the code base that we could apply a more graceful shutdown. Then we could make that graceful shutdown code the core of that PR.

Reply:
https://github.com/rollkit/rollkit/issues/1591 :+1: